h2o-core/src/main/java/water/rapids/BinaryMerge.java

```diff
@@ -357,145 +351,147 @@ private long bsearchLeft(long x, boolean returnLow, long upp)  {
-      if (len > 1) _oneToManyMatch = true;
-      _numRowsInResult += Math.max(1,len) * (lUpp-lLow-1);   // 1 for NA row when _allLeft
-      for (long j = lLow + 1; j < lUpp; j++) {   // usually iterates once only for j=lr, but more than once if there are dup keys in left table
-        // may be a range of left dup'd join-col values, but we need to fetch
-        // each one since the left non-join columns are likely not dup'd and
-        // may be the reason for the cartesian join
-        long t00 = System.nanoTime();
-        // TODO could loop through batches rather than / and % wastefully
-        long globalRowNumber = _leftKO.at8order(j);
-        _timings[17] += (System.nanoTime() - t00)/1e9;
-        t00 = System.nanoTime();
-        int chkIdx = _leftSB._vec.elem2ChunkIdx(globalRowNumber); //binary search in espc
-        _timings[15] += (System.nanoTime() - t00)/1e9;
-        // the key is the same within this left dup range, but still need to fetch left non-join columns
-        _leftKO._perNodeNumRowsToFetch[_leftSB._chunkNode[chkIdx]]++;
-        if (len==0) continue;  // _allLeft must be true if len==0
-
-        // TODO: initial MSB splits should split down to small enough chunk
-        // size - but would that require more passes and if so, how long?  Code
-        // simplification benefits would be welcome!
-        long outLoc = j - (_leftFrom + 1);   // outOffset is 0 here in the standard scaling up high cardinality test
-        // outBatchSize can be different, and larger since known to be 8 bytes
-        // per item, both retFirst and retLen.  (Allowing 8 byte here seems
-        // wasteful, actually.)
-        final int jb2 = (int)(outLoc/_retBatchSize);
-        final int jo2 = (int)(outLoc%_retBatchSize);  // TODO - take outside the loop.  However when we go deep-msb, this'll go away.
-
-        // rLow surrounds row, so +1.  Then another +1 for 1-based
-        // row-number. 0 (default) means nomatch and saves extra set to -1 for
-        // no match.  Could be significant in large edge cases by not needing
-        // to write at all to _ret1st if it has no matches.
-        _ret1st[jb2][jo2] = rLow + 2;
-        _retLen[jb2][jo2] = len;
-      }
-
-      // if we have dup'd left row, we only need to fetch the right rows once
-      // for the first dup.  Those should then be recycled locally later.
-      for (long i=0; i<len; i++) {
-        long loc = rLow+1+i;
-        long t00 = System.nanoTime();
-        // TODO could loop through batches rather than / and % wastefully
+  }
+  
+  private class MidSearchInfo {
+    long _matchIndex;
+    long _lowSearchIndex;
+    long _upperSearchIndex;
+    
+    public MidSearchInfo(long matchI, long lowI, long highI) {
+      _matchIndex = matchI;
+      _lowSearchIndex = lowI;
+      _upperSearchIndex = highI;
+    }
+  }
+  
+  private void populate_ret1st_retLen(int iterLen, int binarySearchLen, long iterIndex, long matchIndex, 
```

Comment:
please try to stick to java naming conventions

Reply:
Good catch.  Fixed.